File: app/src/testOff/java/openfoodfacts/github/scrachx/openfood/network/ProductsAPITest.java
package openfoodfacts.github.scrachx.openfood.network;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.models.ProductState;
import openfoodfacts.github.scrachx.openfood.models.Search;
import openfoodfacts.github.scrachx.openfood.models.entities.SendProduct;
import openfoodfacts.github.scrachx.openfood.network.services.ProductsAPI;
import openfoodfacts.github.scrachx.openfood.utils.Utils;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.jackson.JacksonConverterFactory;

import static com.google.common.truth.Truth.assertThat;

public class ProductsAPITest {
    /**
     * We need to use auth because we use world.openfoodfacts.dev
     */
    private static ProductsAPI devClientWithAuth;
    private static ProductsAPI prodClient;
    private static final String DEV_API = "https://world.openfoodfacts.dev";

    @BeforeClass
    public static void setupClient() {
        OkHttpClient httpClientWithAuth = new OkHttpClient.Builder()
            .addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
            .addInterceptor(chain -> {
                final Request origReq = chain.request();

                Request request = origReq.newBuilder()
                    .header("Authorization", "Basic b2ZmOm9mZg==")
                    .header("Accept", "application/json")
                    .method(origReq.method(), origReq.body()).build();

                return chain.proceed(request);
            })
            .build();

        prodClient = CommonApiManager.getInstance().getProductsApi();

        devClientWithAuth = new Retrofit.Builder()
            .baseUrl(DEV_API)
            .addConverterFactory(JacksonConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .client(httpClientWithAuth)
            .build()
            .create(ProductsAPI.class);
    }

    private static void assertProductsFound(Response<Search> response) {
        assertThat(response).isNotNull();
        assertThat(response.isSuccessful()).isTrue();
        Search search = response.body();

        List<Product> products = search.getProducts();
        assertThat(products).isNotNull();
        assertThat(Integer.parseInt(search.getCount())).isGreaterThan(0);
        assertThat(products.isEmpty()).isFalse();
    }

    private static void assertNoProductsFound(Response<Search> response) {
        assertThat(response).isNotNull();
        assertThat(response.isSuccessful()).isTrue();
        Search search = response.body();
        assertThat(search).isNotNull();

        List<Product> products = search.getProducts();
        assertThat(products.isEmpty()).isTrue();
        assertThat(Integer.parseInt(search.getCount())).isEqualTo(0);
    }

    @Test
    public void byLanguage() throws Exception {
        Response<Search> searchResponse = prodClient.byLanguage("italian").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byLabel() throws Exception {
        Response<Search> searchResponse = prodClient.byLabel("utz-certified").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byCategory() throws Exception {
        Response<Search> searchResponse = prodClient.byCategory("baby-foods").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byState() throws Exception {
        String fieldsToFetchFacets = "brands,product_name,image_small_url,quantity,nutrition_grades_tags";
        Response<Search> searchResponse = prodClient.byState("complete", fieldsToFetchFacets).execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byPackaging() throws Exception {
        Response<Search> searchResponse = prodClient.byPackaging("cardboard").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byBrand() throws Exception {
        Response<Search> searchResponse = prodClient.byBrand("monoprix").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byPurchasePlace() throws Exception {
        Response<Search> searchResponse = prodClient.byPurchasePlace("marseille-5").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byStore() throws Exception {
        Response<Search> searchResponse = prodClient.byStore("super-u").execute();

        assertThat(searchResponse);
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byCountry() throws Exception {
        Response<Search> searchResponse = prodClient.byCountry("france").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byIngredient() throws Exception {
        Response<Search> searchResponse = prodClient.byIngredient("sucre").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void byTrace() throws Exception {
        Response<Search> searchResponse = prodClient.byIngredient("eggs").execute();

        assertThat(searchResponse).isNotNull();
        Search search = searchResponse.body();
        assertThat(search).isNotNull();
        assertThat(search.getProducts()).isNotNull();
    }

    @Test
    public void getProductByTrace_eggs_productsFound() throws Exception {
        Response<Search> response = prodClient.byTrace("eggs").execute();
        assertProductsFound(response);
    }

    @Test
    public void getProductByPackagerCode_emb35069c_productsFound() throws Exception {
        Response<Search> response = prodClient.byPackagerCode("emb-35069c").execute();
        assertProductsFound(response);
    }

    @Test
    public void getProductByNutritionGrade_a_productsFound() throws Exception {
        Response<Search> res = prodClient.byNutritionGrade("a").execute();
        assertProductsFound(res);
    }

    @Test
    public void getProductByCity_Paris_noProductFound() throws Exception {
        Response<Search> response = prodClient.byCity("paris").execute();
        assertNoProductsFound(response);
    }

    @Test
    public void getProductByAdditive_e301_productsFound() throws Exception {
        String fieldsToFetchFacets = "brands,product_name,image_small_url,quantity,nutrition_grades_tags";
        Response<Search> response = prodClient.byAdditive("e301-sodium-ascorbate", fieldsToFetchFacets).execute();
        assertProductsFound(response);
    }

    @Test
    public void getProduct_notFound() throws Exception {
        String barcode = "457457457";
        Response<ProductState> response = prodClient.getProductByBarcode(barcode, "code", Utils.getUserAgent(Utils.HEADER_USER_AGENT_SEARCH)).execute();

        assertThat(response.isSuccessful()).isTrue();

        assertThat(response.body().getStatus()).isEqualTo(0);
        assertThat(response.body().getStatusVerbose()).isEqualTo("product not found");
        assertThat(response.body().getCode()).isEqualTo(barcode);
    }

    @Test
    public void post_product() throws IOException {
        SendProduct product = new SendProduct();

        product.setBarcode("1234567890");
        product.setName("ProductName");
        product.setBrands("productbrand");
        product.setWeight("123");
        product.setWeight_unit("g");
        product.setLang("en");

        Map<String, String> productDetails = new HashMap<String, String>() {{
            put("lang", product.getLang());
            put("product_name", product.getName());
            put("brands", product.getBrands());
            put("quantity", product.getQuantity());
        }};

        ProductState body = devClientWithAuth
            .saveProductSingle(product.getBarcode(), productDetails, OpenFoodAPIClient.getCommentToUpload())
            .blockingGet();

        assertThat(body.getStatus()).isEqualTo(1);
        assertThat(body.getStatusVerbose()).isEqualTo("fields saved");

        String fields = "product_name,brands,brands_tags,quantity";

        Response<ProductState> response = devClientWithAuth.getProductByBarcode(
            product.getBarcode(),
            fields,
            Utils.getUserAgent(Utils.HEADER_USER_AGENT_SEARCH)
        ).execute();

        Product savedProduct = response.body().getProduct();
        assertThat(savedProduct.getProductName()).isEqualTo(product.getName());
        assertThat(savedProduct.getBrands()).isEqualTo(product.getBrands());
        assertThat(savedProduct.getBrandsTags()).contains(product.getBrands());
        assertThat(savedProduct.getQuantity()).isEqualTo(product.getWeight() + " " + product.getWeight_unit());
    }
}

File: app/src/testOff/java/openfoodfacts/github/scrachx/openfood/network/ProductsAPITest.kt
package openfoodfacts.github.scrachx.openfood.network

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.google.common.truth.FailureMetadata
import com.google.common.truth.Subject
import com.google.common.truth.Subject.Factory
import com.google.common.truth.Truth.assertAbout
import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.runBlocking
import okhttp3.Credentials
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import openfoodfacts.github.scrachx.openfood.BuildConfig
import openfoodfacts.github.scrachx.openfood.models.Search
import openfoodfacts.github.scrachx.openfood.models.entities.SendProduct
import openfoodfacts.github.scrachx.openfood.network.ProductsAPITest.SearchSubject.Companion.assertThat
import openfoodfacts.github.scrachx.openfood.network.services.ProductsAPI
import openfoodfacts.github.scrachx.openfood.utils.getUserAgent
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory
import retrofit2.create
import java.time.Duration

class ProductsAPITest {
    @Test
    fun `products by language found`() = runBlocking {
        val search = prodClient.getProductsByLanguage("italian")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by label found`() = runBlocking {
        val search = prodClient.getProductsByLabel("utz-certified")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by category found`() = runBlocking {
        val search = prodClient.getProductsByCategory("baby-foods")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by state found`() = runBlocking {
        val fieldsToFetchFacets = "brands,product_name,image_small_url,quantity,nutrition_grades_tags"
        val search = prodClient.getProductsByState("complete", fieldsToFetchFacets)
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by packaging found`() = runBlocking {
        val search = prodClient.getProductsByPackaging("cardboard")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by brand found`() = runBlocking {
        val search = prodClient.getProductsByBrand("monoprix")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by purchase place found`() = runBlocking {
        val search = prodClient.getProductsByPurchasePlace("marseille-5")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by store found`() = runBlocking {
        val search = prodClient.getProductsByStore("super-u")
        assertThat(search).isNotNull()
        assertThat(search.products).isNotNull()
    }

    @Test
    fun `products by country found`() = runBlocking {
        val search = prodClient.byCountry("france")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by ingredients found`() = runBlocking {
        val search = prodClient.getProductsByIngredient("sucre")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by trace found`() = runBlocking {
        val search = prodClient.getProductsByTrace("eggs")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by trace eggs found`() = runBlocking {
        val response = prodClient.getProductsByTrace("eggs")
        assertThat(response).hasFoundProducts()
    }

    @Test
    fun `product by packager code emb35069c found`() = runBlocking {
        val search = prodClient.getProductsByPackagerCode("emb-35069c")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `product by nutriscore a found`() = runBlocking {
        val search = prodClient.getProductsByNutriScore("a")
        assertThat(search).hasFoundProducts()
    }

    @Test
    fun `products by city paris not found`() = runBlocking {
        val response = prodClient.getProductsByCity("paris")
        assertThat(response).hasFoundNoProducts()
    }

    @ExperimentalCoroutinesApi
    @Test
    fun productByAdditive_e301_productsFound() = runBlocking {
        val fieldsToFetchFacets = "brands,product_name,image_small_url,quantity,nutrition_grades_tags"
        val response = prodClient.getProductsByAdditive("e301-sodium-ascorbate", fieldsToFetchFacets)
        assertThat(response).hasFoundProducts()
    }

    @Test
    fun `product not found`() = runBlocking {
        val barcode = "457457457"
        val state = prodClient.getProductByBarcode(
            barcode,
            "code",
            LC,
            getUserAgent(ApiFields.UserAgents.SEARCH)
        )
        assertThat(state.status).isEqualTo(0)
        assertThat(state.statusVerbose).isEqualTo("product not found")
        assertThat(state.code).isEqualTo(barcode)
    }

    @Test
    fun `post product`() = runBlocking {

        val product = SendProduct().apply {
            barcode = "1234567890"
            name = "test-product-name"
            brands = "test-product-brand"
            weight = "123"
            weightUnit = "g"
            lang = LC
        }

        val productDetails = mapOf<String?, String?>(
            "lang" to product.lang,
            "product_name" to product.name,
            "brands" to product.brands,
            "quantity" to product.quantity
        )


        val body = devClientWithAuth.saveProduct(product.barcode, productDetails, "Automated test")

        assertThat(body.status).isEqualTo(1)
        assertThat(body.statusVerbose).isEqualTo("fields saved")

        val fields = "product_name,brands,brands_tags,quantity"
        val response = devClientWithAuth.getProductByBarcode(
            product.barcode,
            fields,
            LC,
            getUserAgent(ApiFields.UserAgents.SEARCH)
        )

        assertThat(response.product).isNotNull()
        val savedProduct = response.product!!

        assertThat(savedProduct.productName).isEqualTo(product.name)
        assertThat(savedProduct.brands).isEqualTo(product.brands)
        assertThat(savedProduct.brandsTags).contains(product.brands)
        assertThat(savedProduct.quantity).isEqualTo("${product.weight} ${product.weightUnit}")
    }

    companion object {
        const val LC = "en"

        /**
         * We need to use auth because we use world.openfoodfacts.dev
         */
        private lateinit var devClientWithAuth: ProductsAPI
        private lateinit var prodClient: ProductsAPI

        @BeforeAll
        @JvmStatic
        fun setupClient() {
            val httpClientWithAuth = OkHttpClient.Builder()
                .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
                .connectTimeout(Duration.ZERO)
                .readTimeout(Duration.ZERO)
                .addInterceptor {
                    val origReq = it.request()
                    it.proceed(
                        origReq.newBuilder()
                            .header("Authorization", Credentials.basic("off", "off"))
                            .header("Accept", "application/json")
                            .method(origReq.method(), origReq.body())
                            .build()
                    )
                }
                .build()

            prodClient = defaultBuilder()
                .baseUrl(BuildConfig.HOST)
                .build()
                .create()

            devClientWithAuth = defaultBuilder()
                .baseUrl(BuildConfig.TESTING_HOST)
                .client(httpClientWithAuth)
                .build()
                .create()
        }

        private fun defaultBuilder() = Retrofit.Builder()
            .addConverterFactory(JacksonConverterFactory.create(jacksonObjectMapper()))

    }

    class SearchSubject private constructor(
        failureMetadata: FailureMetadata,
        private val actual: Search,
    ) : Subject(failureMetadata, actual) {

        fun hasFoundProducts() {
            check("count").that(actual.count.toInt()).isGreaterThan(0)
            check("products").that(actual.products).isNotEmpty()
        }

        fun hasFoundNoProducts() {
            check("count").that(actual.count.toInt()).isEqualTo(0)
            check("products").that(actual.products).isEmpty()
        }

        companion object {
            private fun searches() = Factory(::SearchSubject)

            fun assertThat(actual: Search): SearchSubject {
                return assertAbout(searches()).that(actual)
            }
        }
    }
}
